index.js

```diff
@@ -96,54 +135,146 @@ original error: ${e}`);
   }
 }
 
-async function exportPodcasts(podcastsDBData) {
+
+function handleSpaces(s) {
+  ret = s;
+  if (argv.nospaces) {
+    ret = s.replaceAll(' ', '_');
+  }
+  return ret;
+}
+
+
+async function buildPodcastDict(fileName, cacheFilesPath, podcastsDBData) {
+  const uuid = fileName.replace(".mp3", "");
+  const dbMeta = podcastsDBData.find((m) => m.zuuid === uuid);
+  const path = `${cacheFilesPath}/${fileName}`;
+  const exportBase = dbMeta?.zcleanedtitle // 1. from apple podcast database
+        ?? (await getMP3MetaTitle(path)) // 2. from mp3 meta data
+        ?? uuid; // 3. fallback to unreadable uuid
+  const podcastName = dbMeta?.zpodcast.replaceAll('/', '_');
+  const exportFileName = sanitize(exportBase.substr(0, fileNameMaxLength));
+
+  const ret = {
+    podcastName: handleSpaces(podcastName),
+    date: dbMeta?.date,
+    fileName,
+    path,
+    uuid,
+    exportFileName: handleSpaces(`${exportFileName}.mp3`)
+  };
+  return ret;
+}
+
+
+function filterPodcasts(podcasts, filepatterns = []) {
+  if (filepatterns.length == 0) {
+    return podcasts;
+  }
+
+  function matchesAny(s) {
+    return filepatterns.some((p) => { return s.indexOf(p) != -1 })
+  }
+
+  return podcasts.filter((p) => {
+    return matchesAny(p.exportFileName) || matchesAny(p.podcastName);
+  });
+}
+
+
+async function exportSingle(podcast, newPath) {
+  copyFileSync(podcast.path, newPath);
+  if (podcast.date && argv.updateutime) {
+    const d = new Date(podcast.date);
+    await fs.utimes(newPath, d, d);
+  }
+}
+
+
+async function exportPodcasts(podcastsDBData, filepatterns = []) {
   const cacheFilesPath = await getPodcastsCacheFilesPath();
   const podcastMP3Files = await getPodcastsCacheMP3Files(cacheFilesPath);
-  const filesWithDBData = podcastMP3Files.map((fileName) => {
-    const uuid = fileName.replace(".mp3", "");
-    const dbMeta = podcastsDBData.find((m) => m.zuuid === uuid);
-    return {
-      fileName,
-      uuid,
-      path: `${cacheFilesPath}/${fileName}`,
-      dbMeta
-    };
-  });
+  const podcasts = await Promise.all(podcastMP3Files.map((fileName) => {
+    return buildPodcastDict(fileName, cacheFilesPath, podcastsDBData);
+  }));
+  let filteredPodcasts = filterPodcasts(podcasts, filepatterns);
+
+  // Weirdly, there are some podcasts that are duplicates ...  i.e.,
```

Comment:
I don't have duplicates, but in theory it can be that app for exampled cached the same podcast 2 times with different uuid, should not really be a problem, even if there are duplicates they will be exported 2 times, if you want to debug it you can export such duplicates with some special name and check whether it's indeed the same podcast episode, but anyway I think that comment can be removed

Reply:
Funnily enough, there were dups in mine!  I didn't look into the podcast db, but when I ran it, the code said "already have X, skipping", even on a fresh export.  Since that looked odd (to me as the user), I left in the check.